third_party/xla_client/pjrt_computation_client.cc

```diff
@@ -524,13 +536,31 @@ PjRtComputationClient::ExecuteReplicated(
   // Required as of cl/518733871
   execute_options.use_major_to_minor_data_layout_for_callbacks = true;
 
+  std::optional<std::vector<PjRtFuture<Status>>> returned_futures(devices.size());
   std::vector<std::vector<std::unique_ptr<PjRtBuffer>>> results =
-      pjrt_computation.executable->Execute(argument_handles, execute_options)
+      pjrt_computation.executable
+          ->Execute(argument_handles, execute_options, returned_futures)
           .value();
 
   std::vector<std::vector<ComputationClient::DataPtr>> data_handles;
   data_handles.reserve(results.size());
   std::vector<size_t> dims(results.size());
+
+  // Grab the shared lock and block the `WaitDeviceOps` until buffer is ready.
+  // Since this is the SPMD code path.
+  // There is no points to grab devices lock for every individual device.
+  auto lock = lock_device_shared(spmd_device_str);
+  TF_VLOG(3) << "ExecuteReplicated grab the lock for " << spmd_device_str;
+  // Signal that `ExecuteReplicated` has completed for one of the devices the
+  // ExecuteReplicatedTime metric. Here, we assume that all devices will finish
+  // execution roughly at the same time, hence only use one of the
+  // returned_futures. Copies the `timed` shared pointer into the lambda.
+  (*returned_futures)[0].OnReady(
+      [timed, lock = std::move(lock)](Status unused) mutable {
+        timed.reset();
```

Comment:
So resetting will break the current timing session and log, is this correct understanding?

Reply:
yea, `timed.reset` will end the metrics session and make it show up in metrics.